File: graph/components/dsu.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * Disjoint Set Union (DSU) / Union-Find
 * Optimizations: Path Compression + Union by Size
 * Time Complexity: O(alpha(N)) ≈ O(1)
 */

struct DSU {
    vector<int> parent;
    vector<int> size;
    int num_sets;
    
    DSU(int n) {
        parent.resize(n + 1);
        size.assign(n + 1, 1);
        num_sets = n;
        // Initialize parent to self
        iota(parent.begin(), parent.end(), 0);
    }
    
    int find(int i) {
        if (parent[i] == i)
            return i;
        return parent[i] = find(parent[i]); // Path compression
    }
    
    bool unite(int i, int j) {
        int root_i = find(i);
        int root_j = find(j);
        
        if (root_i != root_j) {
            // Union by size
            if (size[root_i] < size[root_j])
                swap(root_i, root_j);
            
            parent[root_j] = root_i;
            size[root_i] += size[root_j];
            num_sets--;
            return true;
        }
        return false;
    }
    
    int get_size(int i) {
        return size[find(i)];
    }
};

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    
    int n, m;
    if (!(cin >> n >> m)) return 0;
    
    DSU dsu(n);
    
    for (int i = 0; i < m; i++) {
        int u, v;
        cin >> u >> v;
        if (dsu.unite(u, v)) {
            cout << "Union " << u << " " << v << ": Merged\n";
        } else {
            cout << "Union " << u << " " << v << ": Already in same set\n";
        }
    }
    
    cout << "Number of connected components: " << dsu.num_sets << "\n";
    
    return 0;
}

File: graph/components/dsu.py
import sys

# Fast I/O
input = sys.stdin.readline

class DSU:
    """
    Disjoint Set Union (DSU) / Union-Find with Path Compression and Union by Size
    Time Complexity: O(alpha(N)) which is nearly constant
    """
    def __init__(self, n):
        self.parent = list(range(n + 1))
        self.size = [1] * (n + 1)
        self.num_sets = n
    
    def find(self, i):
        """Find representative of the set containing i with path compression"""
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]
    
    def union(self, i, j):
        """Unions the sets containing i and j"""
        root_i = self.find(i)
        root_j = self.find(j)
        
        if root_i != root_j:
            # Union by size: attach smaller tree to larger tree
            if self.size[root_i] < self.size[root_j]:
                root_i, root_j = root_j, root_i
            
            self.parent[root_j] = root_i
            self.size[root_i] += self.size[root_j]
            self.num_sets -= 1
            return True
        return False
    
    def get_size(self, i):
        return self.size[self.find(i)]

def solve():
    try:
        line1 = input().split()
        if not line1: return
        n, m = map(int, line1)
        
        dsu = DSU(n)
        for _ in range(m):
            u, v = map(int, input().split())
            if dsu.union(u, v):
                print(f"Union {u} {v}: Merged")
            else:
                print(f"Union {u} {v}: Already in same set")
                
        print(f"Number of connected components: {dsu.num_sets}")
            
    except ValueError:
        pass

if __name__ == "__main__":
    solve()
